Updated AuctionHouseSeller to have item class stack control

NathanHandley opened the pull request:
For each item class (consumable, glyph), you can now edit the worldserver.conf to configure the ratio (percent) of random compared to single stack AuctionHouseSeller auction postings.  For example: you can configure it so that all Glyphs are non-stacked, gems are sometimes stacked, reagents are always randomly stacked, etc.  I did this to give a more blizzlike behavior.

**Example 1**
AuctionHouseBot.Class.RandomStackRatio.Consumable = 0
- New consumable auctions posted by AuctionHouseSeller will always be single stack (1).

**Example 2**
AuctionHouseBot.Class.RandomStackRatio.Consumable = 100
- New consumable auctions posted by AuctionHouseSeller will always be of a random stack size (1...n).

**Example 3**
AuctionHouseBot.Class.RandomStackRatio.Consumable = 20
- New consumable auctions posted by AuctionHouseSeller have a 20% chance of being a random stack size (1...n), and an 80% chance of being single stack size (1).

_Note:_ This is my first pull request.  I emulated the coding conventions in the files I changed, including terms (ratio).  Please point out anywhere that I egregiously violated the coding guidelines.

NathanHandley committed:
Updated AuctionHouseSeller to have item class level stack control

For each item class (consumable, glyph), you can now edit the worldserver.conf to

configure the ratio (percent) of random compared to single stack auction postings.

Example 1
AuctionHouseBot.Class.RandomStackRatio.Consumable = 0
- New consumable auctions posted by AuctionHouseSeller will always be single stack.

Example 2
AuctionHouseBot.Class.RandomStackRatio.Consumable = 100
- New consumable auctions posted by AuctionHouseSeller will always be of a random

stack size.

Example 3
AuctionHouseBot.Class.RandomStackRatio.Consumable = 20
- New consumable auctions posted by AuctionHouseSeller have a 20% chance of being a

random stack size, and an 80% chance of being single stack size.

DDuarte commented on src/server/game/AuctionHouseBot/AuctionHouseBotSeller.cpp:
```diff
@@ -952,7 +987,8 @@ void AuctionBotSeller::AddNewAuctions(SellerConfiguration& config)
 
-        uint32 stackCount = urand(1, prototype->GetMaxStackSize());
+        uint32 stackCount = 1;
+        SetStackSizeForItem(prototype, config, stackCount);
```
Make the function return the stackCount instead of using the reference.

DDuarte commented on src/server/game/AuctionHouseBot/AuctionHouseBotSeller.h:
```diff
@@ -39,6 +39,7 @@ struct SellerItemClassInfo
     uint32 MissItems;
     uint32 Quantity;
     uint32 PriceRatio;
+    uint32 RandomStackRatio;
```
Missing initialization in https://github.com/TrinityCore/TrinityCore/blob/3.3.5/src/server/game/AuctionHouseBot/AuctionHouseBotSeller.h#L36

NathanHandley replied:
Thanks-- good catch.  Revised.

DDuarte commented on src/server/worldserver/worldserver.conf.dist:
```diff
@@ -3190,6 +3190,46 @@ AuctionHouseBot.forceIncludeItems = ""
 
 #
+#    AuctionHouseBot.Class.RandomStackRatio.*
+#       Description: Used to determine how often an item of the class will be single or randomly-size stacked
```
I think "an item" should be renamed to a "a stack" since some items cannot stack and therefore this description sounds misleading.

DDuarte commented:
Thanks for the PR :+1: it looks solid. I left some comments above for possible improvements

NathanHandley committed:
Updates in respnose to feedback from DDuarte.  Including:
- Changed SetStackSizeForItem to GetStackSizeForItem in AuctionHouseBotSeller
- Added Initializer to SellerItemClassInfo.RandomStackRatio in AuctionHouseBotSeller
- Updated verbiage around AuctionHouseBot.Class.RandomStackRatio.* to be clearer

NathanHandley commented:
Thank you for the feedback, DDuarte.  I made changes based on all of your comments, as they were good suggestions.  And good catch on the initialization miss.

NathanHandley committed:
Simplified stack size calculation in AuctionBotSeller
